### ISSUE TITLE
numpy.maximum.reduce() doesn't work with object arrays

### ISSUE THREAD

**Issue by mgeier** [numpy/numpy#7698]
I was trying to find the cause for an error in `numpy.linalg.norm()` (see #7575 and #7587) and tracked it to `numpy.maximum.reduce()` (resp. `numpy.minimum.reduce()`).

When I call it with an object array containing another ("normal") array, it doesn't work:

```python
>>> import numpy as np
>>> np.maximum.reduce(np.array([np.array([1, 5]), 3], dtype=object))
...
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

However, when I use the very similar `numpy.add.reduce()`, it works without problems:

```python
>>> np.add.reduce(np.array([np.array([1, 5]), 3], dtype=object))
array([4, 8])
```

I don't grok at all how the low-level implementation of ufuncs and their `reduce()` functions work, so I don't know where to look for a solution.

Any idea how to fix this?

**Comment by pv** [comment]
The reduction on Python objects calls Python's max(a,b) which always 
coerces to bool scalar. So it's likely wontfix.

**Comment by mgeier** [comment]
Thanks for the quick answer, I couldn't have found that out on my own ...

The question is: why?

Why doesn't it call `numpy.max()`?

I assume that `numpy.add.reduce` also calls the corresponding Python operator, but this in turn is pimped by NumPy to handle arrays. AFAIK this is not possible for the built-in `max()` function, therefore it might be more appropriate to call NumPy's max function.

I would like to try what's going to happen when replacing the function call, can you please point me to the place in the sources where `max()` is called?
I assume this happens somewhere in C code, so it would be really nice if you could also tell me how `numpy.max()` can be called with the C API.

**Comment by seberg** [comment]
Numpy never really has a special understanding of arrays inside an array. Arrays of arrays are something I would suggest to avoid if there is an obvious way to avoid it.

The fact that `np.add.reduce` works for these arrays is that `arr + scalar` is well defined and returns a new object. `max(arr, arr)` is undefined, because it can only be resolved on the element level, not the container level (`arr > arr` is undefined). Calling `np.max` is not a completely obvious solution either, since it would break other cases such as `max([1, 2, 3], [1, 2])` which *is* well defined in python.

I also think this is a wontfix, unless there is some more discussion on the mailing list with a few nice arguments for it. If possible, I would suggest to just use the pythons builtin `reduce`/`from functools import reduce`.

Closing for now, feel free to continue the dicussion or discuss on the mailing list how such changes might work.

**Comment by mgeier** [comment]
@seberg Thanks for the explanations!

I'm still curious to try this out. Could somebody please point me to the place where `max()` is called?
And if this happens in C code, how could I call `numpy.max()` instead?

If this works I'll probably continue the discussion on the mailing list, if not, there is no reason to continue the discussion.

**Comment by seberg** [comment]
Heh.... The code is generated by /numpy/core/code_generators/generate_umath.py to call the function `npy_ObjectMax` which is in turn defined in /numpy/core/umath/funcs.inc.src (it is a templated stuff, so look for Max or Object@Kind@).Or you could create a genreation for a new ufunc doing the numpy max.

**Comment by ahaldane** [comment]
Some code that controls this is also in `numpy/core/src/umath/loops.c.src`, where you can see maximum is handled separately from add. (do a search for `maximum` in that file, it's in one of the `repeat` blocks).

By the way, I have an old PR that overhauls how ufuncs work on object arrays in #6320. I've stopped working on it, but maybe it is interesting to you - the diff should highlight the relevant bits of code.

**Comment by mgeier** [comment]
Thanks @ahaldane, your PR #6320 is indeed very interesting!

According to your second point in https://github.com/numpy/numpy/pull/6320#issuecomment-140496431 ("If any of the arguments is a numpy scalar or ndarray then numpy applies the ufunc directly, eg `np.sqrt(elem)`"), this should already work as I want it to, but the check for that doesn't seem to be implemented.

I tried adding this check to `numpy/core/_objectmath.py`:

```python
def maximum(x, y):
    ...
    ...
    if isinstance(x, np.ndarray) or isinstance(y, np.ndarray):
        return np.maximum(x, y)
    ...
    ...
    return max(x, y)
```

... and with that it works perfectly!

```python
>>> import numpy as np
>>> np.maximum.reduce(np.array([np.array([1, 5]), 3], dtype=object))
array([3, 5])
```

Even the case that @seberg mentioned in https://github.com/numpy/numpy/issues/7698#issuecomment-225365541 works as expected:

```python
>>> np.maximum.reduce(np.array(([1, 2, 3], [1, 2]), dtype=object))
[1, 2, 3]
```

Is this check for `np.ndarray` what you had in mind or did I misunderstand your intention?

**Comment by seberg** [comment]
Frankly, I personally think the correct solution would be rather to have an `ndarray` dtype, so that the user can *specify* that they want the element operations be array operations, rather then patching objects to sometimes do this and sometimes that.

**Comment by ahaldane** [comment]
Ah, I see what you're trying to do. I think an alternate fix may be in the C lines involving `if (PyArray_CheckScalar(in1) || ...` to add an extra ` || PyArray_Check(in1)` to catch ndarrays too. Curiously, in one of my last comments in that thread I specifically mention that I didn't allow ndarrays. Maybe I wanted to wait on the decisions for `__numpy_ufunc__` before adding it. 

I tried that change out in a new push to #6320 and rebased.

Also, I agree that a "real" solution doesn't involve object arrays. On the other hand it's pretty easy change in #6320. This way, #6320 makes object arrays more reasonably take care of python numeric types, numpy types, and numpy arrays, and makes no guarantees for anything else. 

Also, note that the rules I wrote in the 2nd comment there are superseded by the 2nd to last comment.

**Comment by ahaldane** [comment]
Hmm it turns out allowing ndarrays in object ufuncs causes a test to fail with that change: 

    >>> a = np.array([np.array([0,2]), 0, 1], dtype='object')
    >>> np.linalg.norm(a, ord=np.inf)   # expected in test to raise ValueError
    array([1, 2])

`linalg.norm` is equivalent to `np.max(np.abs(a))` here. The error is in `np.max`, which used to raise ValueError but with the latest #6320 it returns a result.

So it sounds like your desire for ufuncs to work on nested ndarrays contradicts one unit test.

On the one hand I see why it could be useful and reasonable for `np.max(a)` to return a result here, on the other hand I can see how it could be unexpected and why linalg expects it not too. This is the kind of problem @seberg was talking about finding a "one-size-fits-all" solution.

**Comment by seberg** [comment]
I would be only moderately worried about having to change a single behaviour somewhere (though would have to investigate/understand it first). I think my worry was mostly about incosistencies with arrays vs. array-likes, since such an extra logic can only ever cover arrays, but not general array-likes.

That does not make it impossible, just should think about it a bit more and before the end get a discussion on the list, etc.

**Comment by mgeier** [comment]
Thanks @ahaldane for updating #6320, now it works great for my use case!

I've committed the now failing unit test myself in #7587 a few weeks ago (specifically faa6cf3bcf2a49a4b165a0d17cf349f622ec0b2f).
The test cases check for the *current* behavior, not for the *desired* behavior!
This was probably a silly thing to do ...

If you want, you can use this commit with updated tests: https://github.com/mgeier/numpy/commit/5d9f8b0bab2dfc5288a75bc9089fcc4e57bdea86.

Sadly, some of those tests still don't pass, but that may have to be fixed in the implementation of `numpy.linalg()`.

**Comment by seberg** [comment]
Documenting current behaviour in tests is good in my opinion. You could
only add a comment for someone stumbling on it to reduce the fear of
changing it ;).
